### PR TITLE
[S390-Port]: Remove IC Safepoint

### PR DESCRIPTION
**Pull request by offamitkumar**
Update: With current code changes only `CastNullCheckDroppingsTest.java` test is failing.

contains s390x changes for ic-safepoint removal port. Build for slowdebug, fastdebug & release are working. But few of additional test failure are seen in tier-1 testing. 

```file
CastNullCheckDroppingsTest.java
SortingDeoptimizationTest.java
java/lang/StrictMath/ExhaustingTests.java 
java/math/BigInteger/LargeValueExceptions.java 
java/util/Random/RandomCanaryPi.java 
java/util/Spliterator/SpliteratorCollisions.java 
java/util/stream/test/org/openjdk/tests/java/util/SplittableRandomTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/CollectorsTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/CountLargeTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/FlatMapOpTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/InfiniteStreamWithLimitOpTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/MapOpTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/SliceOpTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/StreamLinkTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/StreamSpliteratorTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/ToArrayOpTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/WhileOpTest.java 
java/util/stream/test/org/openjdk/tests/java/util/stream/mapMultiOpTest.java 
java/util/zip/ZipFile/TestZipFileEncodings.java 
```

These changes needs to applied on top of https://github.com/openjdk/jdk/compare/master...fisk:jdk:remove-ic-safepoints-v2. 

Changes in diff file: 
[ic_safepoint_change.patch](https://github.com/openjdk/jdk/files/13668385/ic_safepoint_change.patch)

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17100/head:pull/17100` \
`$ git checkout pull/17100`

Update a local copy of the PR: \
`$ git checkout pull/17100` \
`$ git pull https://git.openjdk.org/jdk.git pull/17100/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17100`

View PR using the GUI difftool: \
`$ git pr show -t 17100`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17100.diff">https://git.openjdk.org/jdk/pull/17100.diff</a>

</details>

### CODE REVIEW

**Comment by offamitkumar** [comment]
Hi @TheRealMDoerr, @RealLucy  Can I go ahead with this patch for s390 ?

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp [hunk 1]
```diff
@@ -2150,6 +2157,40 @@ void MacroAssembler::call_VM_leaf_base(address entry_point) {
   call_VM_leaf_base(entry_point, allow_relocation);
 }
 
+int MacroAssembler::ic_check_size() {
+  return 30 + (ImplicitNullChecks ? 0 : 6);
```
Comment (TheRealMDoerr) [inline src/hotspot/cpu/s390/macroAssembler_s390.cpp:2161]: Did you check if the size is correct and the alignment works as intended in the -XX:+PrintAssembly output with and without -XX:-ImplicitNullChecks?

**Review by TheRealMDoerr — COMMENTED** [review]
Looks good to me. Please improve the assertion.

File: src/hotspot/cpu/s390/macroAssembler_s390.cpp (hunk 1 above)
Comment (offamitkumar) [reply src/hotspot/cpu/s390/macroAssembler_s390.cpp:2161]: Hi Martin, 

This is output with flag: 
```asm
[Entry Point]
  # {method} {0x000003ff500007d0} 'hashCode' '()I' in 'java/lang/Object'
  #           [sp+0xd0]  (sp of caller)
  0x000003ff8cc30980:   nopr
  0x000003ff8cc30982:   nopr
  0x000003ff8cc30984:   nopr
  0x000003ff8cc30986:   nopr
  0x000003ff8cc30988:   nopr
  0x000003ff8cc3098a:   nopr
  0x000003ff8cc3098c:   nopr
  0x000003ff8cc3098e:   nopr
  0x000003ff8cc30990:   nopr
  0x000003ff8cc30992:   nopr
  0x000003ff8cc30994:   nopr
  0x000003ff8cc30996:   nopr
  0x000003ff8cc30998:   nopr
  0x000003ff8cc3099a:   nopr
  0x000003ff8cc3099c:   cgije   %r2,0,0x000003ff8cc309b2
  0x000003ff8cc309a2:   llgf    %r1,8(%r2)
  0x000003ff8cc309a8:   cg  %r1,8(%r9)
  0x000003ff8cc309ae:   je  0x000003ff8cc309c0
  0x000003ff8cc309b2:   iihf    %r1,1023                    ;const 0x3ff8c51b8c0 | 4396105709760 |   4.396105709760000e+12
  0x000003ff8cc309b8:   iilf    %r1,2354165952
  0x000003ff8cc309be:   br  %r1
```

And this is without ImplicitNullCheck flag: 
```asm
[Entry Point]
  # {method} {0x000003ff540007d0} 'hashCode' '()I' in 'java/lang/Object'
  #           [sp+0xd0]  (sp of caller)
  0x000003ff94c30980:   nopr
  0x000003ff94c30982:   nopr
  0x000003ff94c30984:   nopr
  0x000003ff94c30986:   nopr
  0x000003ff94c30988:   nopr
  0x000003ff94c3098a:   nopr
  0x000003ff94c3098c:   nopr
  0x000003ff94c3098e:   nopr
  0x000003ff94c30990:   nopr
  0x000003ff94c30992:   nopr
  0x000003ff94c30994:   nopr
  0x000003ff94c30996:   nopr
  0x000003ff94c30998:   nopr
  0x000003ff94c3099a:   nopr
  0x000003ff94c3099c:   nopr
  0x000003ff94c3099e:   nopr
  0x000003ff94c309a0:   nopr
  0x000003ff94c309a2:   llgf    %r1,8(%r2)
  0x000003ff94c309a8:   cg  %r1,8(%r9)
  0x000003ff94c309ae:   je  0x000003ff94c309c0
  0x000003ff94c309b2:   iihf    %r1,1023                    ;const 0x3ff9451b8c0 | 4396239927488 |   4.396239927488000e+12
  0x000003ff94c309b8:   iilf    %r1,2488383680
  0x000003ff94c309be:   br  %r1
[Verified Entry Point]
```

Also with latest patch from Erik, `CastNullCheckDroppingsTest.java` test is also passing.

**Comment by TheRealMDoerr** [comment]
I think it's good. I recommend checking performance using some benchmark. Performance was very poor on PPC64 before this fix: https://github.com/fisk/jdk/commit/5aed08975dd9a9cb00acd4b4f4c071dfb962ac69
I didn't see any regression afterwards.

**Comment by offamitkumar** [comment]
>I think it's good. I recommend checking performance using some benchmark.

Would you please suggest which benchmark should I use here ( maybe you can tell me which benchmark you used for testing) ?

**Comment by TheRealMDoerr** [comment]
I have used SPEC benchmarks, but you can try any other one if you don't have a license. I don't remember which one you have used last time. I guess you will see a big difference if the fix is missing.

**Comment by offamitkumar** [comment]
> I have used SPEC benchmarks, but you can try any other one if you don't have a license. 

I guess we have license, So I'll try to access it.

> I guess you will see a big difference if the fix is missing.

Does that mean most of the benchmark will show the regression ? My understanding was that, different benchmarks are there to test different functionality.

**Comment by TheRealMDoerr** [comment]
If such fundamental things are broken, basically every benchmark should notice.

**Comment by offamitkumar** [comment]
Hi Martin, 
I used renaissance benchmark and here is the result I got: 
with patch: 
```
Benchmark                                  Mode  Cnt      Score      Error  Units
o.r.actors.JmhAkkaUct.run                    ss   35   2705.773 ±  292.097  ms/op
o.r.actors.JmhReactors.run                   ss   15  20903.519 ± 1931.106  ms/op
o.r.jdk.concurrent.JmhFjKmeans.run           ss   40   4153.656 ±  386.497  ms/op
o.r.jdk.concurrent.JmhFutureGenetic.run      ss   65   3184.723 ±  125.662  ms/op
o.r.jdk.streams.JmhMnemonics.run             ss   25   8046.079 ±  314.166  ms/op
o.r.jdk.streams.JmhParMnemonics.run          ss   25   6528.264 ±  213.007  ms/op
o.r.jdk.streams.JmhScrabble.run              ss   65    277.482 ±   47.013  ms/op
o.r.neo4j.JmhNeo4jAnalytics.run              ss   30   2843.985 ±  103.831  ms/op
o.r.rx.JmhRxScrabble.run                     ss  105    217.778 ±   17.829  ms/op
o.r.scala.dotty.JmhDotty.run                 ss   65   1043.688 ±   67.138  ms/op
o.r.scala.sat.JmhScalaDoku.run               ss   30   7670.109 ±  756.502  ms/op
o.r.scala.stdlib.JmhScalaKmeans.run          ss   65    312.690 ±   15.646  ms/op
o.r.scala.stm.JmhPhilosophers.run            ss   40  10197.090 ±  716.933  ms/op
o.r.scala.stm.JmhScalaStmBench7.run          ss   80   4152.354 ±  123.389  ms/op
o.r.twitter.finagle.JmhFinagleChirper.run    ss  115  23160.800 ±  391.440  ms/op
o.r.twitter.finagle.JmhFinagleHttp.run       ss   20  10253.295 ± 1136.701  ms/op
```

without the patch: 
```
Benchmark                                  Mode  Cnt      Score      Error  Units
o.r.actors.JmhAkkaUct.run                    ss   35   2880.461 ±  296.531  ms/op
o.r.actors.JmhReactors.run                   ss   15  21792.433 ± 2346.775  ms/op
o.r.jdk.concurrent.JmhFjKmeans.run           ss   40   4259.036 ±  498.852  ms/op
o.r.jdk.concurrent.JmhFutureGenetic.run      ss   65   3199.238 ±  118.804  ms/op
o.r.jdk.streams.JmhMnemonics.run             ss   25   8488.351 ±  282.508  ms/op
o.r.jdk.streams.JmhParMnemonics.run          ss   25   7459.727 ±  202.225  ms/op
o.r.jdk.streams.JmhScrabble.run              ss   65    274.635 ±   45.646  ms/op
o.r.neo4j.JmhNeo4jAnalytics.run              ss   30   2959.414 ±   79.198  ms/op
o.r.rx.JmhRxScrabble.run                     ss  105    233.800 ±   19.422  ms/op
o.r.scala.dotty.JmhDotty.run                 ss   65   1053.025 ±   87.842  ms/op
o.r.scala.sat.JmhScalaDoku.run               ss   30   6547.708 ± 1425.924  ms/op
o.r.scala.stdlib.JmhScalaKmeans.run          ss   65    288.766 ±   16.716  ms/op
o.r.scala.stm.JmhPhilosophers.run            ss   40  10191.924 ±  564.472  ms/op
o.r.scala.stm.JmhScalaStmBench7.run          ss   80   4414.132 ±  163.167  ms/op
o.r.twitter.finagle.JmhFinagleChirper.run    ss  115  23875.359 ±  491.453  ms/op
o.r.twitter.finagle.JmhFinagleHttp.run       ss   20  12543.643 ±  600.994  ms/op
```


This is single-run result, and machine is shared-resource. I was only able to run it once (with & without patch) as it takes more than 6 hours to finish.

**Comment by offamitkumar** [comment]
Hi @TheRealMDoerr, 
Should I close this one if everything looks okay to you ?

**Comment by TheRealMDoerr** [comment]
Your s390 code is already included in the official PR. So, this one can get closed.

**Comment by offamitkumar** [comment]
Closing in favour of https://github.com/openjdk/jdk/pull/17495

**Pull request closed without merging**